Add DEAL_II_WITH_TET_SUPPORT flag

peterrum opened the pull request:
... and a test folder (with a dummy test).

As suggested by @tamiko :)

tamiko commented on include/deal.II/base/config.h.in:
```diff
@@ -38,6 +38,7 @@
 #cmakedefine DEAL_II_WITH_ADOLC
 #cmakedefine DEAL_II_WITH_ARPACK
 #cmakedefine DEAL_II_WITH_ASSIMP
+#cmakedefine DEAL_II_WITH_TET_MESH_SUPPORT
```
my OCD demands it: Would you mind sorting alphabetically? :-)

peterrum replied:
Done!

tjhei commented:
this will also be used for simulations not only for tet "meshes" right? I would call it DEAL_II_WITH_TET_SUPPORT

peterrum commented:
/rebuild

peterrum commented:
> this will also be used for simulations not only for tet "meshes" right? I would call it DEAL_II_WITH_TET_SUPPORT

Personally, I would like to keep `DEAL_II_WITH_TET_MESH_SUPPORT` since I mainly intend to use it to guard the `Triangulation` class. Many functions/classes that have been or will be generalized with TETs in mind will be part of the default configuration.

But let's  make a quick poll (or collect some alternative suggestions): a *laught* for `DEAL_II_WITH_TET_MESH_SUPPORT` and *rocket* for `DEAL_II_WITH_TET_SUPPORT`.

peterrum commented:
> But let's make a quick poll

Any other opinions?

tamiko commented:
@peterrum Let's just go with `WITH_TET_SUPPORT` instead. We can revisit our naming choice before the next release.

peterrum commented:
OK! I have mad the name change!

tjhei commented on tests/tet/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.12)
+INCLUDE(../setup_testsubproject.cmake)
+PROJECT(testsuite CXX)
+DEAL_II_PICKUP_TESTS()
```
I would do
```
IF (DEAL_II_WITH_TET_SUPPORT)
  DEAL_II_PICKUP_TESTS()
ENDIF()
```

tjhei replied:
but this can be a followup when you actually add useful tests here.

peterrum replied:
I thought about this. But, I find it too restrictive, since there will be tests which will work with the default configuration.

bangerth commented:
I'm late to the discussion, but doesn't the flag also guard the use of *triangles*, and not just tetrahedra? Shouldn't it be `DEAL_II_WITH_SIMPLEX_SUPPORT`?

bangerth commented on cmake/setup_cached_variables.cmake:
```diff
@@ -372,6 +372,12 @@ OPTION(DEAL_II_WITH_64BIT_INDICES
 LIST(APPEND DEAL_II_FEATURES 64BIT_INDICES)
 
+OPTION(DEAL_II_WITH_TET_SUPPORT
+  "If set to ON, triangulations with triangle and tetrahedron cells are supported in addition to hex-only triangulations."
```
quadrilateral- and hexahedra-only triangulations.

bangerth commented on tests/tet/dummy.with_tet_support=off.output:
```diff
@@ -0,0 +1,2 @@
+
+DEAL::deal.II is configured without TET support.
```
I don't think you need the two variants -- the test is simply never picked up when support is `off`. In other words, if the test for whatever reason is run without support (and consequently outputs `without`), then the test pickup logic is borken, the test fails the comparison to the one available output file, and that is exactly what should happen.

bangerth commented:
You probably also want to set the flag to `ON` in `doc/doxygen/options.dox.in`.

peterrum commented:
I have removed the test output and changed the documentation. I would like to stick with the `DEAL_II_WITH_TET_SUPPORT` for now. I think enabling the flag in `doc/doxygen/options.dox.in` is a bit premature. Let's enable it once we have something to show!

bangerth commented:
On 6/24/20 2:49 PM, Peter Munch wrote:
> I would like to stick with the |DEAL_II_WITH_TET_SUPPORT| for now.

The boat has shipped, but I still wanted to ask: Why use an inaccurate name if 
we know that it is inaccurate and have a better name? (Namely, 
|DEAL_II_WITH_SIMPLEX_SUPPORT|.)

drwells commented:
I wasn't paying attention to this thread but I think simplex is preferable to tet.

masterleinad commented:
There is still time until the next release. :slightly_smiling_face:

peterrum commented:
> Why use an inaccurate name if we know that it is inaccurate and have a better name?

It might be inaccurate, but I would argue that more people are familiar with the term *tet*. In engineering, many people don't even know the term *hypercube*. 

I would suggest that we postpone the discussion about the name of the flag until before the release when we know what features we have actually implemented until that time. With that knowledge, we can appropriately select a name which describes these features.
